### PR TITLE
T5222: Add load-balancing for web traffic

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Add load-balancing for web traffic
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T5222

## Component(s) name

load-balancing
### Without SSL termination, based on domain name
```
set load-balancing reverse-proxy server https backend 'bk-default'
set load-balancing reverse-proxy server https description 'bind app listen on 443 port'
set load-balancing reverse-proxy server https mode 'tcp'
set load-balancing reverse-proxy server https port '443'

set load-balancing reverse-proxy server https rule 10 domain-name 'node1.example.com'
set load-balancing reverse-proxy server https rule 10 set server 'bk-api-01'
set load-balancing reverse-proxy server https rule 20 domain-name 'node2.example.com'
set load-balancing reverse-proxy server https rule 20 set server 'bk-api-02'

set load-balancing reverse-proxy backend bk-api-01 description 'My API-1'
set load-balancing reverse-proxy backend bk-api-01 mode 'tcp'
set load-balancing reverse-proxy backend bk-api-01 server api01 address '127.0.0.1'
set load-balancing reverse-proxy backend bk-api-01 server api01 port '8481'
set load-balancing reverse-proxy backend bk-api-02 description 'My API-2'
set load-balancing reverse-proxy backend bk-api-02 mode 'tcp'
set load-balancing reverse-proxy backend bk-api-02 server api01 address '127.0.0.2'
set load-balancing reverse-proxy backend bk-api-02 server api01 port '8882'

set load-balancing reverse-proxy backend bk-default description 'Default backend'
set load-balancing reverse-proxy backend bk-default mode 'tcp'
set load-balancing reverse-proxy backend bk-default server sr01 address '127.0.0.3'
set load-balancing reverse-proxy backend bk-default server sr01 port '8888'

```
Check haproxy.cfg
```
vyos@r14# sudo haproxy -f /run/haproxy/haproxy.cfg -c
Configuration file is valid
[edit]
vyos@r14# 
```
haproxy.cfg
```
vyos@r14# cat /run/haproxy/haproxy.cfg 
# Generated by ${vyos_conf_scripts_dir}/load-balancing-haproxy.py

global
    log /dev/log local0
    log /dev/log local1 notice
    chroot /var/lib/haproxy
    stats socket /run/haproxy/admin.sock mode 660 level admin
    stats timeout 30s
    user haproxy
    group haproxy
    daemon


    # Default SSL material locations
    ca-base /etc/ssl/certs
    crt-base /etc/ssl/private

    # https://ssl-config.mozilla.org/#server=haproxy&version=2.6.12-1&config=intermediate&openssl=3.0.8-1&guideline=5.6
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
    ssl-default-bind-options force-tlsv13

defaults
    log     global
    mode    http
    option  dontlognull
    timeout connect 10s
    timeout client  50s
    timeout server  50s
    errorfile 400 /etc/haproxy/errors/400.http
    errorfile 403 /etc/haproxy/errors/403.http
    errorfile 408 /etc/haproxy/errors/408.http
    errorfile 500 /etc/haproxy/errors/500.http
    errorfile 502 /etc/haproxy/errors/502.http
    errorfile 503 /etc/haproxy/errors/503.http
    errorfile 504 /etc/haproxy/errors/504.http

# Frontend
frontend https
    bind *:443 
    mode tcp
    # rule 10
    acl 10 hdr(host) -i node1.example.com
    use_backend bk-api-01 if 10
    # rule 20
    acl 20 hdr(host) -i node2.example.com
    use_backend bk-api-02 if 20
    default_backend bk-default


# Backend
backend bk-api-01
    balance roundrobin
    mode tcp
    server api01 127.0.0.1:8481  

backend bk-api-02
    balance roundrobin
    mode tcp
    server api01 127.0.0.2:8882  

backend bk-default
    balance roundrobin
    mode tcp
    server sr01 127.0.0.3:8888  


```
### Terminate SSL
redirect ` '/.well-known/foo'` to `/certs/my/`
redirect ` '/.well-known/bar'` to `/certs/ok/`
```
set load-balancing reverse-proxy server http description 'Force redirect to HTTPS'
set load-balancing reverse-proxy server http port '80'
set load-balancing reverse-proxy server http redirect-http-to-https

set load-balancing reverse-proxy server https backend 'bk-default'
set load-balancing reverse-proxy server https description 'bind app listen on 443 port'
set load-balancing reverse-proxy server https mode 'http'
set load-balancing reverse-proxy server https port '443'
set load-balancing reverse-proxy server https rule 10 set redirect-location '/certs/my/'
set load-balancing reverse-proxy server https rule 10 url-path exact '/.well-known/foo'
set load-balancing reverse-proxy server https rule 20 set redirect-location '/certs/ok/'
set load-balancing reverse-proxy server https rule 20 url-path end '/mail'
set load-balancing reverse-proxy server https rule 20 url-path exact '/.well-known/bar'

set load-balancing reverse-proxy server https ssl certificate 'cert'

set load-balancing reverse-proxy backend bk-default description 'Default backend'
set load-balancing reverse-proxy backend bk-default mode 'http'
set load-balancing reverse-proxy backend bk-default server sr01 address '127.0.0.3'
set load-balancing reverse-proxy backend bk-default server sr01 port '8888'

```
Check haproxy.cfg:
```
vyos@r14# sudo haproxy -f /run/haproxy/haproxy.cfg -c
Configuration file is valid
[edit]
vyos@r14# 

```
Cat haproxy.cfg
```
# Generated by ${vyos_conf_scripts_dir}/load-balancing-haproxy.py

global
    log /dev/log local0
    log /dev/log local1 notice
    chroot /var/lib/haproxy
    stats socket /run/haproxy/admin.sock mode 660 level admin
    stats timeout 30s
    user haproxy
    group haproxy
    daemon


    # Default SSL material locations
    ca-base /etc/ssl/certs
    crt-base /etc/ssl/private

    # https://ssl-config.mozilla.org/#server=haproxy&version=2.6.12-1&config=intermediate&openssl=3.0.8-1&guideline=5.6
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
    ssl-default-bind-options force-tlsv13

defaults
    log     global
    mode    http
    option  dontlognull
    timeout connect 10s
    timeout client  50s
    timeout server  50s
    errorfile 400 /etc/haproxy/errors/400.http
    errorfile 403 /etc/haproxy/errors/403.http
    errorfile 408 /etc/haproxy/errors/408.http
    errorfile 500 /etc/haproxy/errors/500.http
    errorfile 502 /etc/haproxy/errors/502.http
    errorfile 503 /etc/haproxy/errors/503.http
    errorfile 504 /etc/haproxy/errors/504.http

# Frontend
frontend http
    bind *:80 
    http-request redirect scheme https unless { ssl_fc }

frontend https
    bind *:443 ssl crt /run/haproxy/cert.pem
    mode http
    # rule 10
    acl 10 path  /.well-known/foo
    http-request redirect location /certs/my/ code 301 if 10
    # rule 20
    acl 20 path -i -m end /mail
    acl 20 path  /.well-known/bar
    http-request redirect location /certs/ok/ code 301 if 20
    default_backend bk-default


# Backend
backend bk-default
    balance roundrobin
    option forwardfor
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    mode http
    server sr01 127.0.0.3:8888  


```

### L4 balancing
```
set load-balancing reverse-proxy server front01 backend 'bk01'
set load-balancing reverse-proxy server front01 mode 'tcp'
set load-balancing reverse-proxy server front01 port '8888'

set load-balancing reverse-proxy backend bk01 mode 'tcp'
set load-balancing reverse-proxy backend bk01 server bk01 address '192.168.122.11'
set load-balancing reverse-proxy backend bk01 server bk01 port '8881'
set load-balancing reverse-proxy backend bk01 server bk02 address '192.168.122.11'
set load-balancing reverse-proxy backend bk01 server bk02 port '8882'

set load-balancing reverse-proxy global-parameters tls tls-version-min '1.2'

```
Check haproxy.cfg
```
vyos@r14# sudo haproxy -f /run/haproxy/haproxy.cfg -c
Configuration file is valid
[edit]
vyos@r14# 
```
haproxy.cfg
```
# Generated by ${vyos_conf_scripts_dir}/load-balancing-haproxy.py

global
    log /dev/log local0
    log /dev/log local1 notice
    chroot /var/lib/haproxy
    stats socket /run/haproxy/admin.sock mode 660 level admin
    stats timeout 30s
    user haproxy
    group haproxy
    daemon


    # Default SSL material locations
    ca-base /etc/ssl/certs
    crt-base /etc/ssl/private

    # https://ssl-config.mozilla.org/#server=haproxy&version=2.6.12-1&config=intermediate&openssl=3.0.8-1&guideline=5.6
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
    ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets

defaults
    log     global
    mode    http
    option  dontlognull
    timeout connect 10s
    timeout client  50s
    timeout server  50s
    errorfile 400 /etc/haproxy/errors/400.http
    errorfile 403 /etc/haproxy/errors/403.http
    errorfile 408 /etc/haproxy/errors/408.http
    errorfile 500 /etc/haproxy/errors/500.http
    errorfile 502 /etc/haproxy/errors/502.http
    errorfile 503 /etc/haproxy/errors/503.http
    errorfile 504 /etc/haproxy/errors/504.http

# Frontend
frontend front01
    bind *:8888 
    mode tcp
    default_backend bk01


# Backend
backend bk01
    balance roundrobin
    mode tcp
    server bk01 192.168.122.11:8881  
    server bk02 192.168.122.11:8882  

```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: data/templates/load-balancing/haproxy.cfg.j2 [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    # Default SSL material locations
+    ca-base /etc/ssl/certs
+    crt-base /etc/ssl/private
+
+{%     if global_parameters.tls.ssl_bind_ciphers is vyos_defined %}
+    ssl-default-bind-ciphers {{ global_parameters.tls.ssl_bind_ciphers | join(':') | upper }}
+{%     else %}
+    # https://ssl-config.mozilla.org/#server=haproxy&version=2.6.12-1&config=intermediate&openssl=3.0.8-1&guideline=5.6
+    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
```
Comment (c-po) [inline data/templates/load-balancing/haproxy.cfg.j2]: Please use `<defaultValue>` for defaults. Multiple defaults can be generated using whitespace separation.

File: data/templates/load-balancing/haproxy.cfg.j2 [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    errorfile 500 /etc/haproxy/errors/500.http
+    errorfile 502 /etc/haproxy/errors/502.http
+    errorfile 503 /etc/haproxy/errors/503.http
+    errorfile 504 /etc/haproxy/errors/504.http
+
+# Frontend
+{% if server is vyos_defined %}
+{%     for front, front_config in server.items() %}
+frontend {{ front }}
+    bind {{ front_config.listen_address if front_config.listen_address if vyos_defined else '' }}:{{ front_config.port }}
```
Comment (c-po) [inline data/templates/load-balancing/haproxy.cfg.j2]: Do we really need the `else` here?

File: data/templates/load-balancing/haproxy.cfg.j2 [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+{%         endif %}
+    option forwardfor
+    http-request set-header X-Forwarded-Port %[dst_port]
+    http-request add-header X-Forwarded-Proto https if { ssl_fc }
+{%         if back_config.mode is vyos_defined %}
+    mode {{ back_config.mode }}
+{%         endif %}
+{%         if back_config.server is vyos_defined %}
+{%             for server, server_config in back_config.server.items() %}
+    server {{ server }} {{ server_config.address }}:{{ server_config.port }} {{ 'check' if server_config.check is vyos_defined else '' }}
```
Comment (c-po) [inline data/templates/load-balancing/haproxy.cfg.j2]: Do we really need the `else` here?

File: data/templates/load-balancing/override_haproxy.conf.j2 [hunk 4]
```diff
@@ -0,0 +1,13 @@
+[Unit]
```
Comment (c-po) [inline data/templates/load-balancing/override_haproxy.conf.j2:2]: Maybe direclty add support to bind haproxy to a VRF?

File: interface-definitions/load-balancing-haproxy.xml.in [hunk 5]
```diff
@@ -0,0 +1,200 @@
+<?xml version="1.0"?>
+<interfaceDefinition>
+  <node name="load-balancing">
+    <children>
+      <node name="reverse-proxy" owner="${vyos_conf_scripts_dir}/load-balancing-haproxy.py">
+        <properties>
+          <help>Configure reverse-proxy load-balancing</help>
```
Comment (c-po) [inline interface-definitions/load-balancing-haproxy.xml.in]: IMHO `load-balancing` is redundant here.

File: interface-definitions/load-balancing-haproxy.xml.in [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        </properties>
+        <children>
+          <tagNode name="server">
+            <properties>
+              <help>Frontend name</help>
```
Comment (c-po) [inline interface-definitions/load-balancing-haproxy.xml.in:12]: Should we limit server name to be alphanumeric only?

File: interface-definitions/load-balancing-haproxy.xml.in [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            </properties>
+            <children>
+              <leafNode name="backend">
+                <properties>
+                  <help>Backend member</help>
```
Comment (c-po) [inline interface-definitions/load-balancing-haproxy.xml.in:21]: Should we limit backend server names to be alphanumeric only? It's a "virtual" identifier inside haproxy conf. I guess we do not wan't to have fancy characters like äöüß

File: interface-definitions/load-balancing-haproxy.xml.in [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+                </properties>
+              </leafNode>
+              #include <include/listen-address.xml.i>
+              #include <include/haproxy/mode.xml.i>
+              #include <include/port-number.xml.i>
+            </children>
+          </tagNode>
+          <tagNode name="backend">
+            <properties>
+              <help>Backend name</help>
```
Comment (c-po) [inline interface-definitions/load-balancing-haproxy.xml.in:59]: Same constaint for alphanumeric as above

File: interface-definitions/load-balancing-haproxy.xml.in [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+                  </leafNode>
+                  #include <include/port-number.xml.i>
+                </children>
+              </tagNode>
+              #include <include/haproxy/timeout.xml.i>
+            </children>
+          </tagNode>
+          <node name="global-parameters">
+            <properties>
+              <help>global-parameters</help>
```
Comment (c-po) [inline interface-definitions/load-balancing-haproxy.xml.in]: Can we improve this help string?

File: interface-definitions/load-balancing-haproxy.xml.in [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+                        <description>dhe-rsa-aes256-gcm-sha384</description>
+                      </valueHelp>
+                      <constraint>
+                        <regex>(ecdhe-ecdsa-aes128-gcm-sha256|ecdhe-rsa-aes128-gcm-sha256|ecdhe-ecdsa-aes256-gcm-sha384|ecdhe-rsa-aes256-gcm-sha384|ecdhe-ecdsa-chacha20-poly1305|ecdhe-rsa-chacha20-poly1305|dhe-rsa-aes128-gcm-sha256|dhe-rsa-aes256-gcm-sha384)
+                        </regex>
+                      </constraint>
+                      <multi/>
+                    </properties>
+                  </leafNode>
+                  <leafNode name="tls-version-min">
```
Comment (c-po) [inline interface-definitions/load-balancing-haproxy.xml.in:216]: We should be good internet citizens and default this to 1.3

File: src/conf_mode/load-balancing-haproxy.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    if not lb:
+        return None
+
+    if 'backend' not in lb or 'server' not in lb:
+        raise ConfigError(f'"server" and "backend" must be configured!')
+
+    for front, front_config in lb['server'].items():
+        if 'port' not in front_config:
+            raise ConfigError(f'"{front} server port" must be configured!')
+
```
Comment (c-po) [inline src/conf_mode/load-balancing-haproxy.py:93]: Should we check for bind ports already used?

**Review by c-po — CHANGES_REQUESTED** [review]

File: data/templates/load-balancing/haproxy.cfg.j2 (hunk 1 above)
Comment (sever-sever) [reply data/templates/load-balancing/haproxy.cfg.j2]: Done!

File: data/templates/load-balancing/haproxy.cfg.j2 (hunk 2 above)
Comment (sever-sever) [reply data/templates/load-balancing/haproxy.cfg.j2]: Fixed!

File: data/templates/load-balancing/haproxy.cfg.j2 (hunk 3 above)
Comment (sever-sever) [reply data/templates/load-balancing/haproxy.cfg.j2]: Fixed

File: interface-definitions/load-balancing-haproxy.xml.in (hunk 5 above)
Comment (sever-sever) [reply interface-definitions/load-balancing-haproxy.xml.in]: Done!

File: interface-definitions/load-balancing-haproxy.xml.in (hunk 7 above)
Comment (sever-sever) [reply interface-definitions/load-balancing-haproxy.xml.in:21]: Done!

File: interface-definitions/load-balancing-haproxy.xml.in (hunk 8 above)
Comment (sever-sever) [reply interface-definitions/load-balancing-haproxy.xml.in:59]: Done!

File: interface-definitions/load-balancing-haproxy.xml.in (hunk 9 above)
Comment (sever-sever) [reply interface-definitions/load-balancing-haproxy.xml.in]: Done!

File: interface-definitions/load-balancing-haproxy.xml.in (hunk 10 above)
Comment (sever-sever) [reply interface-definitions/load-balancing-haproxy.xml.in:216]: Set default 1.3, done!

File: src/conf_mode/load-balancing-haproxy.py (hunk 11 above)
Comment (sever-sever) [reply src/conf_mode/load-balancing-haproxy.py:93]: Done!

File: data/templates/load-balancing/haproxy.cfg.j2 (hunk 2 above)
Comment (sever-sever) [reply data/templates/load-balancing/haproxy.cfg.j2]: > Do we really need the `else` here?


There are 2 examples of configuration, in one case it used '*' in another empty ''
```
bind *:12345
bind :12345
```
Not sure which option feet better

File: data/templates/load-balancing/override_haproxy.conf.j2 (hunk 4 above)
Comment (sever-sever) [reply data/templates/load-balancing/override_haproxy.conf.j2:2]: Done!
```
vyos@r14# run show conf com | match vrf
set load-balancing reverse-proxy vrf 'foo'
set vrf name foo table '1010'
[edit]
vyos@r14# 
[edit]
vyos@r14# sudo ip vrf pids foo
17806  haproxy
17808  haproxy
[edit]
vyos@r14# 

```

File: data/templates/load-balancing/haproxy.cfg.j2 (hunk 2 above)
Comment (c-po) [reply data/templates/load-balancing/haproxy.cfg.j2]: `*` expresses to `any` IP address - we might wan't to make it a CLI completion helper option. So user selects either explicitly `any` or a distinct IP address

**Comment by c-po** [comment]
Please also add an explicit dependency to `debian/control` to import haproxy. As of now this is only a side-load of ocserv.

File: data/templates/load-balancing/haproxy.cfg.j2 (hunk 2 above)
Comment (sever-sever) [reply data/templates/load-balancing/haproxy.cfg.j2]: I mean for template, should we remove **else** or use **else '*'**
```
bind {{ front_config.listen_address if front_config.listen_address if vyos_defined }}:{{ front_config.port }} {{ ssl_front }}
```
or
```
bind {{ front_config.listen_address if front_config.listen_address if vyos_defined else '*' }}:{{ front_config.port }} {{ ssl_front }}
```

**Comment by sever-sever** [comment]
> Please also add an explicit dependency to `debian/control` to import haproxy. As of now this is only a side-load of ocserv.

@c-po Done!

File: data/templates/load-balancing/haproxy.cfg.j2 (hunk 2 above)
Comment (sever-sever) [reply data/templates/load-balancing/haproxy.cfg.j2]: Update. 
I set `*` by default is listen address is not defined

**Review by c-po — APPROVED** [review]

**Pull request merged**